File: src/SlimShader/Chunks/Shex/Tokens/TessellatorDomainDeclarationToken.cs
using SlimShader.Chunks.Common;
using SlimShader.Util;

namespace SlimShader.Chunks.Shex.Tokens
{
	/// <summary>
	/// Hull Shader Declaration Phase: Tessellator Domain
	///
	/// OpcodeToken0:
	///
	/// [10:00] D3D11_SB_OPCODE_DCL_TESS_DOMAIN
	/// [12:11] Domain
	/// [23:13] Ignored, 0
	/// [30:24] Instruction length in DWORDs including the opcode token. == 1
	/// [31]    0 normally. 1 if extended operand definition, meaning next DWORD
	///         contains extended operand description.  This dcl is currently not
	///         extended.
	/// </summary>
	public class TessellatorDomainDeclarationToken : DeclarationToken
	{
		public TessellatorDomain Domain { get; private set; }

		public static TessellatorDomainDeclarationToken Parse(BytecodeReader reader)
		{
			var token0 = reader.ReadUInt32();
			return new TessellatorDomainDeclarationToken
			{
				Domain = token0.DecodeValue<TessellatorDomain>(11, 12)
			};
		}

		public override string ToString()
		{
			return string.Format("{0} {1}", TypeDescription, Domain.GetDescription(ChunkType.Shex));
		}
	}
}

File: src/cpp/Source/Chunks/Shex/Tokens/TessellatorDomainDeclarationToken.cpp
#include "PCH.h"
#include "TessellatorDomainDeclarationToken.h"

#include "Decoder.h"

using namespace std;
using namespace SlimShader;

shared_ptr<TessellatorDomainDeclarationToken> TessellatorDomainDeclarationToken::Parse(BytecodeReader& reader)
{
	auto token0 = reader.ReadUInt32();
	auto result = shared_ptr<TessellatorDomainDeclarationToken>(new TessellatorDomainDeclarationToken());
	result->_domain = DecodeValue<TessellatorDomain>(token0, 11, 12);
	return result;
};

TessellatorDomain TessellatorDomainDeclarationToken::GetDomain() const { return _domain; }

void TessellatorDomainDeclarationToken::Print(std::ostream& out) const
{
	out << GetTypeDescription() << " " << ToStringShex(_domain);
};